### ISSUE TITLE
Implement DP-ASSM block (Dual-Path Attention + SSM)

### ISSUE THREAD

**Issue by z-a-f** [maida-ai/Research#3]
Combine windowed local attention with a global state-space path, then fuse via a learnable gate.

## Description

Create a transformer block that:

1. Applies **causal windowed attention** over the most recent `W` tokens for precise local reasoning.
2. Maintains a compact **SSM state** to summarize long-range context with linear cost.
3. **Fuses** the two paths with a learned gate per token.

## Implementation Suggestions

* File: `efficient_longctx/blocks/dpassm.py`
* Skeleton:
  ```python
  import torch
  import torch.nn as nn

  class DPASSMBlock(nn.Module):
      def __init__(self, d_model, n_heads, window_size, ssm_state_dim, dropout=0.1, n_probes=0):
          super().__init__()
          # TODO: self.attn = FlashAttention-based windowed attention
          # TODO: self.ssm = simple SSM cell (see notes below)
          self.probes = nn.Parameter(torch.randn(n_probes, d_model)) if n_probes > 0 else None
          self.gate = nn.Linear(d_model, 1)
          self.proj = nn.Linear(d_model, d_model)
          self.dropout = nn.Dropout(dropout)

      def forward(self, x, ssm_state=None):
          # x: (B, T, d_model)
          y_attn = self._local_attn(x)                     # implement causal windowed attention
          y_ssm, new_ssm_state = self._ssm_path(x, ssm_state)  # recurrent update over T
          gate = torch.sigmoid(self.gate(x))
          y = gate * y_attn + (1 - gate) * y_ssm
          return self.proj(self.dropout(y)), new_ssm_state
  ```

* Local attention:
  * Use FlashAttention-2 causal kernel over sliding window `W`.
  * Mask tokens outside the window.
* SSM path (simple starter):
  * Implement a small gated 1D convolution or a diagonal-plus-low-rank recurrence:
    * `s_t = A * s_{t-1} + B * x_t`, `y_ssm_t = C * s_t` with learnable `A, B, C` (start with diagonal `A`).
* Add `tests/test_dpassm.py` to check shapes, gradients, and state continuity across two forward calls.

## Acceptance Criteria

* [ ] `DPASSMBlock` forward pass runs on CUDA with random input.
* [ ] Attention uses a causal window; outside-window tokens are not attended.
* [ ] SSM state is updated and returned; continuity verified by test; two-chunk forward ≈ one-shot forward (unit test).
* [ ] Feature-wise gate mixes the two paths; gradients flow end-to-end.
* [ ] Residual + MLP with Pre-LN implemented.
* [ ] All tests in `tests/test_dpassm.py` pass locally.

---

## What we’re building (at a glance)

Input tokens $x \in \mathbb{R}^{B\times T\times d_{model}}$ go through **two parallel paths**:

1. **Local attention path** (precise, short-term):

* Each token only attends to the previous $W$ tokens (including itself), **causally**.

2. **SSM path** (compact, long-term):

* A small recurrent state $s_t \in \mathbb{R}^{d_s}$ is updated one step at a time and produces a per-token output summarizing all history.

We then **fuse** the two path outputs per token with a **learned gate** and apply standard **residual + MLP** like a Transformer block.

## Math (clear, exact shapes)

Let `d = d_model`, `h = n_heads`, `d_h = d/h`, `d_s = ssm_state_dim`. For token index `t = 1..T`:

### 1) Local, windowed causal attention

Project queries/keys/values (per head):

* $Q_t = x_t W_Q \in \mathbb{R}^{h \times d_h}$
* $K_i = x_i W_K \in \mathbb{R}^{h \times d_h}$
* $V_i = x_i W_V \in \mathbb{R}^{h \times d_h}$

For position `t`, only indices $i \in [\max(1, t-W+1), t]$ are visible. Scores per head:

* $\text{score}_{t,i} = \frac{ \langle Q_t, K_i \rangle }{\sqrt{d_h}}$

Masked softmax over the allowed window:

* $p_{t,i} = \text{softmax}_i(\text{score}_{t,i})$

Head outputs:

* $O_t^{(head)} = \sum_i p_{t,i}, V_i \in \mathbb{R}^{d_h}$

Concatenate heads and project:

* $y^{attn}_t = \text{Concat}_h(O_t^{(head)}) W_O \in \mathbb{R}^{d}$

### 2) State-space (SSM) path (simple linear recurrent cell)

Let the **state** be $s_t \in \mathbb{R}^{d_s}$. Parameters:

* $A \in \mathbb{R}^{d_s}$ (diagonal coefficients), treated as elementwise $\alpha$
* $B \in \mathbb{R}^{d_s \times d}$
* $C \in \mathbb{R}^{d \times d_s}$

Recurrence (per token):

* $s_t = \alpha \odot s_{t-1} + B x_t$   (elementwise $\odot$; $s_0 = 0$ or passed from previous chunk)
* $y^{ssm}_t = C, s_t \in \mathbb{R}^{d}$

> Optional nonlinearity after state or output (e.g., `tanh(s_t)` or `SiLU(C s_t)`) can be added later; start linear for clarity.

### 3) Fusion gate (dimension-wise gate recommended)

Compute a **gate per feature** so the model can choose attention vs SSM **per channel**:

* $g_t = \sigma(W_g x_t + b_g) \in \mathbb{R}^{d}$  (sigmoid in [0,1] elementwise)

Fuse:

* $y_t = g_t \odot y^{attn}_t + (1 - g_t) \odot y^{ssm}_t \in \mathbb{R}^{d}$

### 4) Residual & MLP (Pre-LN block)

Use **Pre-LayerNorm** everywhere (stable, standard):

1. `x1 = x + Dropout(y)`
2. `x2 = x1 + Dropout(MLP(LN(x1)))`, where MLP is:

   * $\text{MLP}(u) = W_2 , \text{GELU}(W_1 u)$, with hidden dim `4d` (configurable).

---

## Reference code stubs (safe to copy and fill in)

> **Heads up:** These are **minimal** and emphasize correctness. Optimize later.

### Helper: windowed causal mask

```python
def make_window_mask(T: int, W: int, device, dtype):
    # allow j in [i-W+1, i] and j <= i
    i = torch.arange(T, device=device)
    j = torch.arange(T, device=device)
    I, J = torch.meshgrid(i, j, indexing="ij")
    causal = (J <= I)
    window = (I - J) < W
    allow = causal & window
    mask = torch.zeros((T, T), device=device, dtype=dtype)
    mask = mask.masked_fill(~allow, float("-inf"))
    return mask  # (T, T)
```

### Local attention (using SDPA)

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

def split_heads(x, h):
    B, T, D = x.shape
    d_h = D // h
    return x.view(B, T, h, d_h).transpose(1, 2)  # (B,h,T,d_h)

def merge_heads(x):
    B, h, T, d_h = x.shape
    return x.transpose(1, 2).contiguous().view(B, T, h * d_h)

class WindowedAttention(nn.Module):
    def __init__(self, d_model, n_heads, window_size, dropout=0.0):
        super().__init__()
        self.d = d_model
        self.h = n_heads
        self.Wq = nn.Linear(d_model, d_model, bias=False)
        self.Wk = nn.Linear(d_model, d_model, bias=False)
        self.Wv = nn.Linear(d_model, d_model, bias=False)
        self.Wo = nn.Linear(d_model, d_model, bias=False)
        self.drop = nn.Dropout(dropout)
        self.W = window_size

    def forward(self, x):
        B, T, D = x.shape
        q = split_heads(self.Wq(x), self.h)  # (B,h,T,d_h)
        k = split_heads(self.Wk(x), self.h)
        v = split_heads(self.Wv(x), self.h)

        # SDPA expects (B*h, T, d_h)
        q = q.reshape(B * self.h, T, -1)
        k = k.reshape(B * self.h, T, -1)
        v = v.reshape(B * self.h, T, -1)

        mask = make_window_mask(T, self.W, device=x.device, dtype=torch.float32)  # float mask
        # broadcast to (B*h, T, T)
        mask = mask.unsqueeze(0).expand(B * self.h, -1, -1)

        y = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, is_causal=True)  # (B*h, T, d_h)
        y = y.reshape(B, self.h, T, -1)
        y = merge_heads(y)  # (B,T,D)
        return self.Wo(self.drop(y))
```

### SSM cell (simple, stable)

```python
class SimpleSSM(nn.Module):
    def __init__(self, d_model, d_state):
        super().__init__()
        self.A = nn.Parameter(torch.zeros(d_state))  # will squash to (-1,1)
        self.B = nn.Linear(d_model, d_state, bias=False)
        self.C = nn.Linear(d_state, d_model, bias=False)

    def forward(self, x, s0=None):
        """
        x: (B,T,d_model)
        s0: (B,d_state) or None
        returns: y_ssm: (B,T,d_model), sT: (B,d_state)
        """
        B, T, D = x.shape
        d_state = self.A.shape[0]
        if s0 is None:
            s = x.new_zeros(B, d_state)
        else:
            s = s0
        y = []
        alpha = torch.tanh(self.A)  # keep in (-1,1)
        for t in range(T):
            s = alpha * s + self.B(x[:, t, :])
            y.append(self.C(s))
        y = torch.stack(y, dim=1)  # (B,T,D)
        return y, s
```

### DP-ASSM block (wire together)

```python
class DPASSMBlock(nn.Module):
    def __init__(self, d_model, n_heads, window_size, ssm_state_dim, dropout=0.1):
        super().__init__()
        self.ln1 = nn.LayerNorm(d_model)
        self.ln2 = nn.LayerNorm(d_model)
        self.attn = WindowedAttention(d_model, n_heads, window_size, dropout=dropout)
        self.ssm = SimpleSSM(d_model, ssm_state_dim)
        self.gate = nn.Linear(d_model, d_model)  # feature-wise gate
        self.drop = nn.Dropout(dropout)
        self.mlp = nn.Sequential(
            nn.Linear(d_model, 4 * d_model),
            nn.GELU(),
            nn.Linear(4 * d_model, d_model),
        )

    def forward(self, x, ssm_state=None):
        """
        x: (B,T,d_model)
        ssm_state: (B,d_state) or None
        returns: (B,T,d_model), (B,d_state)
        """
        u = self.ln1(x)
        y_attn = self.attn(u)                 # (B,T,d)
        y_ssm, new_state = self.ssm(u, ssm_state)  # (B,T,d), (B,d_s)

        g = torch.sigmoid(self.gate(u))       # (B,T,d)
        y = g * y_attn + (1.0 - g) * y_ssm

        x = x + self.drop(y)                  # residual 1
        x = x + self.drop(self.mlp(self.ln2(x)))  # residual 2
        return x, new_state
```

---

## Tests (put in `tests/test_dpassm.py`)

1. **Shapes & dtype**

* Create block with small dims (e.g., `d=32, h=4, W=8, d_s=16`), random `x` on CPU and GPU (if available).
* Assert output shape `(B,T,d)` and state shape `(B,d_s)`.

2. **Causality & windowing**

* Create a tiny `T=6, W=3` case with crafted `x` so that changing `x[:, t-4, :]` **does not** change `y[:, t, :]` (outside window), while changing `x[:, t-2, :]` **does**.
* Easiest: run forward twice, flip a token far outside the window, assert `torch.allclose` for the target time slice.

3. **SSM continuity across chunks**

* Run full forward on `x_full` → `(y_full, s_full)`.
* Split `x_full` into two parts: `x_a, x_b`.

  * Run forward on `x_a` → `(y_a, s_a)`.
  * Run forward on `x_b` with `ssm_state=s_a` → `(y_b, s_b)`.
* Check `torch.allclose(torch.cat([y_a,y_b],dim=1), y_full, atol=1e-5, rtol=1e-4)` and `s_b ≈ s_full`.

4. **Gradients**

* Do a simple loss, e.g., `y.mean()` on GPU; call `loss.backward()`; ensure no error and some parameters have non-zero grads.

5. **CPU fallback**

* Ensure tests run on CPU if CUDA isn’t present (skip GPU-specific asserts with `pytest.importorskip("torch.cuda")` / `pytest.mark.skipif`).

---

## Performance & correctness notes

* The **mask** is $T\times T$ and may be heavy for very large `T`. For Week-1 correctness and unit tests it’s fine. Later, we’ll:

  * Use **chunked windows** to avoid full masks, or
  * Switch to FlashAttention-2 **varlen** APIs for memory-efficient windowed attention.
* The **SSM** is implemented as a simple for-loop. That’s fine for now and very clear. Later, we can:

  * Vectorize with scan-like kernels or blockwise parallelization,
  * Add **gated updates** or a small nonlinearity if quality needs it.
